### PR TITLE
Lazily load the s3 library

### PR DESCRIPTION
**Pull request by bdurette**
Based on existing work in the lazyload branch, this change defers the loading of the aws library functions until they are needed. Data from our production environments show this will save ~250ms per page load.

### CODE REVIEW

File: inc/class-s3-uploads-stream-wrapper.php [hunk 1]
```diff
@@ -2,16 +2,23 @@
...
 	 * Register the 's3://' stream wrapper
 	 *
 	 * @param S3Client $client Client to use with the stream wrapper
 	 */
-	public static function register( Aws\S3\S3Client $client) {
-		stream_wrapper_register( 's3', __CLASS__, STREAM_IS_URL );
-		static::$client = $client;
-	}
+	public static function register_streamwrapper( $s3_uploads )
+	{
```
Comment (rmccue) [inline inc/class-s3-uploads-stream-wrapper.php]: Watch the coding standards here; WP coding standards dictate the brace on the same line as the declaration :)

File: inc/class-s3-uploads-stream-wrapper.php [hunk 2]
```diff
@@ -2,16 +2,23 @@
...
 	 */
-	public static function register( Aws\S3\S3Client $client) {
-		stream_wrapper_register( 's3', __CLASS__, STREAM_IS_URL );
-		static::$client = $client;
-	}
+	public static function register_streamwrapper( $s3_uploads )
+	{	
+		static::$s3_uploads = $s3_uploads;
+
+		if (in_array('s3', stream_get_wrappers())) {
```
Comment (rmccue) [inline inc/class-s3-uploads-stream-wrapper.php]: These need spaces in the parentheses, per coding standards.

File: inc/class-s3-uploads-stream-wrapper.php [hunk 3]
```diff
@@ -164,4 +171,8 @@ public function url_stat( $path, $flags ) {
 	public function stream_metadata( $path, $option, $value ) {
 		// not implemented
 	}
+
+	public function __construct() {
+		static::$client = static::$s3_uploads->s3();
+	}
```
Comment (rmccue) [inline inc/class-s3-uploads-stream-wrapper.php]: Can we move this to be the first method on the class?

**Comment by rmccue** [comment]
Thanks for the patch! Quick coding standards review, will let @joehoyle review proper. :)

**Comment by bdurette** [comment]
Cleaned up the code to meet the coding standards. Thanks for the review!

**Comment by joehoyle** [comment]
Ran tests locally, as running on an external remote currently isn't supported. Thanks @bdurette!

**Pull request merged**